test/lit/passes/cast-refining.wast

```diff
@@ -0,0 +1,1039 @@
+  (type $B (struct_subtype $A))
+
+  ;; YESTNH:      (type $C (struct_subtype  $B))
+  ;; NO_TNH:      (type $C (struct_subtype  $B))
+  (type $C (struct_subtype $B))
+
+  ;; YESTNH:      (type $anyref_=>_none (func (param anyref)))
+
+  ;; YESTNH:      (func $new (type $anyref_=>_none) (param $x anyref)
+  ;; YESTNH-NEXT:  (drop
+  ;; YESTNH-NEXT:   (struct.new_default $C)
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT: )
+  ;; NO_TNH:      (func $new (type $anyref_=>_none) (param $x anyref)
+  ;; NO_TNH-NEXT:  (drop
+  ;; NO_TNH-NEXT:   (struct.new_default $C)
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT: )
+  (func $new (param $x anyref)
+    (drop
+      (struct.new $C)
+    )
+  )
+
+  ;; YESTNH:      (func $ref.cast (type $anyref_=>_none) (param $x anyref)
+  ;; YESTNH-NEXT:  (drop
+  ;; YESTNH-NEXT:   (ref.cast $C
+  ;; YESTNH-NEXT:    (local.get $x)
+  ;; YESTNH-NEXT:   )
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT:  (drop
+  ;; YESTNH-NEXT:   (ref.cast $C
+  ;; YESTNH-NEXT:    (local.get $x)
+  ;; YESTNH-NEXT:   )
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT:  (drop
+  ;; YESTNH-NEXT:   (ref.cast $C
+  ;; YESTNH-NEXT:    (local.get $x)
+  ;; YESTNH-NEXT:   )
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT: )
+  ;; NO_TNH:      (func $ref.cast (type $anyref_=>_none) (param $x anyref)
+  ;; NO_TNH-NEXT:  (drop
+  ;; NO_TNH-NEXT:   (ref.cast $A
+  ;; NO_TNH-NEXT:    (local.get $x)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT:  (drop
+  ;; NO_TNH-NEXT:   (ref.cast $B
+  ;; NO_TNH-NEXT:    (local.get $x)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT:  (drop
+  ;; NO_TNH-NEXT:   (ref.cast $C
+  ;; NO_TNH-NEXT:    (local.get $x)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT: )
+  (func $ref.cast (param $x anyref)
+    (drop
+      (ref.cast $A     ;; This can be $C.
+        (local.get $x)
+      )
+    )
+    (drop
+      (ref.cast $B     ;; This can also be $C.
+        (local.get $x)
+      )
+    )
+    (drop
+      (ref.cast $C
+        (local.get $x)
+      )
+    )
+  )
+)
+
+;; More testing for cases where no types or subtypes are created. No type is
+;; created here.
+(module
+  ;; YESTNH:      (type $anyref_=>_none (func (param anyref)))
+
+  ;; YESTNH:      (type $A (struct ))
+  ;; NO_TNH:      (type $A (struct ))
+  (type $A (struct))
+
+  ;; YESTNH:      (type $B (struct_subtype  $A))
+  ;; NO_TNH:      (type $anyref_=>_none (func (param anyref)))
+
+  ;; NO_TNH:      (type $B (struct_subtype  $A))
+  (type $B (struct_subtype $A))
+
+  ;; YESTNH:      (type $C1 (struct_subtype  $B))
+  ;; NO_TNH:      (type $C1 (struct_subtype  $B))
+  (type $C1 (struct_subtype $B))
+
+  ;; YESTNH:      (type $C2 (struct_subtype  $B))
+  ;; NO_TNH:      (type $C2 (struct_subtype  $B))
+  (type $C2 (struct_subtype $B))
+
+  ;; YESTNH:      (func $ref.cast (type $anyref_=>_none) (param $x anyref)
+  ;; YESTNH-NEXT:  (drop
+  ;; YESTNH-NEXT:   (block
+  ;; YESTNH-NEXT:    (drop
+  ;; YESTNH-NEXT:     (ref.cast $A
```

Comment:
I wonder if it would be worth updating these casts to be casts to bottom type in addition to sequencing an `unreachable` after them. That would remove the type use and further empower e.g. MergeTypes without having to run an Vacuum in between.

Reply:
Hmm, I'm not sure it's worth it. CastTypes will definitely be followed up with Vacuum in the normal pipeline (it's in pre global opts, so it runs before all the func opts).